### PR TITLE
[pallet-revive] Use an EIP-684 collision check for contract creation

### LINKED ISSUE #13128: pallet-revive: creation-collision guard is account-type, not EIP-684 — CREATE onto DelegatedEOA is accepted
## Summary

`ContractInfo::new` rejects only `AccountInfo::is_contract` (`AccountType::Contract`). An address that is already `AccountType::DelegatedEOA` is accepted for CREATE/CREATE2. `insert_contract` then overwrites the nested `contract_info` while keeping `DelegatedEOA` + `payer`.

**No practical attack path is claimed.** Joining a 7702 authority (`keccak(ecdsa_recover(...))[12..]`) with a CREATE/CREATE2 destination (`keccak(...)[12..]`) is an address collision; a keyless Nick-style deployer cannot sign an authorization. This is a correctness / EIP-684 gap and a code-identity inconsistency, not a bounty-style exploit.

Related prior art: #12639 (collision guard is only `is_contract`; prescribe EIP-684 nonce-based check), #12804 (same family).

## What goes wrong if the state is forced

When CREATE lands on a forced `DelegatedEOA` address (tests can construct this; production cannot reasonably):

1. **Code-identity spoofing (strongest observable):** `eth_getCode` / delegation indicator still reports `0xef0100||T`, while execution uses the CREATE `code_hash`. EXTCODEHASH checks `get_delegation_target` first (`exec.rs` ~2469–2472); EXTCODESIZE checks `is_delegated` first (`exec.rs` ~2502).
2. **Deposit / trie bookkeeping:** `trie_id` swap orphans prior `storage_base_deposit` / byte / item deposits and the delegated child trie; clear/re-delegate refunds against the *recorded* (stale) `payer` and CREATE-sized base. Permanent entry deposit after clear is already documented accepted design in `prdoc/pr_12229.prdoc` — not the novel part.
3. **DeletionQueue** may not reclaim the orphaned trie entries in that forced path.

## Sites that need the stronger predicate

| Site | Today |
|------|--------|
| `exec.rs` ~1222 instantiate via `ContractInfo::new` | only `is_contract` |
| `exec.rs` ~1147 precompile/`ContractInfo::new` path | same |
| `exec.rs` ~2099 `terminate_if_same_tx` | lacks the `is_delegated` guard that `terminate_caller` has |

Production block for *delegate onto contract* is `is_contract` at `eip7702.rs` ~143 (not the `debug_assert!` at `storage.rs` ~370–373, which is compiled out in release).

Note: under `PGasDeposit`, `PendingDepositCleanup` (`storage.rs` ~462) can also refuse instantiate when stale `NativeDepositOf` rows remain — do not ignore that guard when reasoning about deposit outcomes.

## Suggested fix

Replace the creation-collision check with an EIP-684-style predicate, e.g. reject if any of:

- nonzero nonce,
- code or delegation indicator present,
- non-empty contract/delegation trie / non-EOA account type,

rather than a `DelegatedEOA`-only special case. Align `terminate_if_same_tx` with the delegated terminate guard.

## Refs

- PR #12229 / merge `4d63d92fa5e0a31c03385fa01f76c65be7ed83f4`
- `substrate/frame/revive/src/storage.rs` — `ContractInfo::new`, `insert_contract` DelegatedEOA arm
- `substrate/frame/revive/src/exec.rs` — instantiate, EXTCODE*, terminate paths
- `substrate/frame/revive/src/evm/eip7702.rs` — authority recovery / `is_contract` on auth
- #12639, #12804

Happy to turn this into a PR if useful.

### PR DESCRIPTION
**Pull request by Nathy-bajo**
Fixes #13128 by making CREATE and CREATE2 reject any address with a nonzero nonce, code, or an EIP-7702 delegation entry, and by stopping `SELFDESTRUCT` from scheduling a delegated EOA for deletion while still moving its balance.

### CODE REVIEW

File: substrate/frame/revive/src/lib.rs
Comment (0xRVE) [inline substrate/frame/revive/src/lib.rs:1]: Fix the doc for `DuplicateContract` error variant

File: substrate/frame/revive/src/storage.rs [hunk 1]
```diff
@@ -189,6 +189,17 @@ impl<T: Config> AccountInfo<T> {
...
+	/// EIP-684: deploying to `address` collides if it has a nonce, code, or a delegation entry.
+	///
+	/// A cleared `DelegatedEOA` still counts: its child trie and deposit accounting outlive the
+	/// delegation and would be orphaned by a contract taking over the entry.
+	pub fn is_create_collision(address: &H160) -> bool {
+		<AccountInfoOf<T>>::get(address)
+			.is_some_and(|info| !matches!(info.account_type, AccountType::EOA)) ||
+			!frame_system::Pallet::<T>::account_nonce(&T::AddressMapper::to_account_id(address))
+				.is_zero()
+	}
```
Comment (0xRVE) [inline substrate/frame/revive/src/storage.rs:201]: this adds a read and needs to be benchmarked.

File: substrate/frame/revive/src/exec/tests.rs [hunk 2]
```diff
@@ -1437,6 +1437,65 @@ fn termination_from_instantiate_succeeds() {
...
+				Pallet::<Test>::convert_native_to_evm(min_balance * 100),
+				vec![],
+				&ExecConfig::new_substrate_tx(),
+			));
+
+			let address = instantiated.borrow().unwrap();
+			assert_eq!(
+				crate::storage::AccountInfo::<Test>::get_delegation_target(&address),
+				Some(CHARLIE_ADDR)
+			);
```
Comment (0xRVE) [inline substrate/frame/revive/src/exec/tests.rs:1495]: add `assert_eq!(Pallet::<Test>::evm_balance(&address), U256::zero());`

**Review by 0xRVE — APPROVED** [review]
Approved, pls fix the comments

**Comment by Nathy-bajo** [comment]
> Approved, pls fix the comments

Thanks, all done

**Review by 0xRVE — APPROVED** [review]

File: substrate/frame/revive/src/storage.rs [hunk 3]
```diff
@@ -444,17 +455,28 @@ impl<T: Config> AccountInfo<T> {
...
-	/// in storage.
+	/// This returns an `Err` if deploying to `address` collides with an existing account, see
+	/// [`AccountInfo::is_create_collision`].
 	pub fn new(
 		address: &H160,
 		nonce: T::Nonce,
 		code_hash: sp_core::H256,
 	) -> Result<Self, DispatchError> {
-		if <AccountInfo<T>>::is_contract(address) {
+		if <AccountInfo<T>>::is_create_collision(address) {
```
Comment (athei) [inline substrate/frame/revive/src/storage.rs:465]: EIP-684 also consumes the creator's nonce. The nonce is incremented before the collision check and stays incremented when the create fails; geth sets `nonce + 1` first and only then returns `ErrContractAddressCollision`. Here the error comes out of `new_frame` before `run` bumps the caller's nonce (`exec.rs:1442-1446`), so a colliding CREATE/CREATE2 leaves the factory's nonce unchanged, and a factory whose next CREATE address is taken hits the same address on every retry.

The same bump is also lost when the constructor reverts, because it runs inside the child's `with_transaction` and is rolled back with it. In Ethereum a reverted CREATE still consumes the nonce.

Please move the caller's nonce bump out of `do_transaction` into the `Instantiate` arm of `new_frame`, right after the address is derived and before `ContractInfo::new`, with the same `bump_nonce || !is_first_frame` condition. Then assert the factory's nonce in the Solidity test below, after a collision and after a reverting constructor.

File: substrate/frame/revive/src/storage.rs [hunk 4]
```diff
@@ -189,6 +189,17 @@ impl<T: Config> AccountInfo<T> {
 		matches!(info.account_type, AccountType::Contract(_))
 	}
 
+	/// EIP-684: deploying to `address` collides if it has a nonce, code, or a delegation entry.
+	///
+	/// A cleared `DelegatedEOA` still counts: its child trie and deposit accounting outlive the
+	/// delegation and would be orphaned by a contract taking over the entry.
+	pub fn is_create_collision(address: &H160) -> bool {
```
Comment (athei) [inline substrate/frame/revive/src/storage.rs:196]: This makes a leftover nonce permanent. Every contract starts at nonce 1 (`exec.rs:1440`), and `do_terminate` relies on the account being reaped to clear it. If the contract holds a sufficient asset when it is terminated (for example USDT on Asset Hub through the assets precompile), `dec_providers` keeps the System account alive with sufficients > 0 and nonce >= 1 while `AccountInfoOf` is removed. On master a CREATE2 redeploy to that address works, and it is the only way to move those assets out again. With this check it fails with `DuplicateContract` forever. In Ethereum the whole account including the nonce is deleted.

Please reset the nonce in `do_terminate` when the account survives the final transfer. Nobody controls the fallback account of a contract address short of an address collision, so there is no replay concern. And add a test: deploy, keep the account alive with a sufficient (an asset, or `inc_sufficients` if the mock has no assets pallet), terminate, redeploy with the same salt.

File: substrate/frame/revive/src/storage.rs [hunk 5]
```diff
@@ -189,6 +189,17 @@ impl<T: Config> AccountInfo<T> {
...
+		<AccountInfoOf<T>>::get(address)
+			.is_some_and(|info| !matches!(info.account_type, AccountType::EOA)) ||
+			!frame_system::Pallet::<T>::account_nonce(&T::AddressMapper::to_account_id(address))
```
Comment (athei) [inline substrate/frame/revive/src/storage.rs:199]: `to_account_id` reads `OriginalAccount`. That read is new on every instantiate and not in the current weights: `seal_instantiate` and `evm_instantiate` list no `OriginalAccount` access, and the one in `instantiate` is the caller's mapping check. `to_account_id` is the right account to check here, since it resolves to the account behind the address, so please keep it and re-run the instantiate benchmarks (`instantiate`, `instantiate_with_code`, `eth_instantiate_with_code`, `seal_instantiate`, `evm_instantiate`) so the read is charged. That also settles the benchmarking question from the earlier thread.

File: substrate/frame/revive/src/exec.rs [hunk 6]
```diff
@@ -1317,7 +1317,7 @@ where
 			self.exec_config,
 		)? {
 			// EIP-684: an in-construction address is not in `AccountInfoOf` yet, so the
-			// `is_contract` guard in `ContractInfo::new` misses this re-entrant collision.
+			// collision check in `ContractInfo::new` can miss this re-entrant collision.
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:1320]: This is no longer true. `run` bumps the new contract's nonce to 1 before any constructor code runs (`exec.rs:1440`), so a re-entrant CREATE2 to the in-construction address is already rejected by `ContractInfo::new` inside `new_frame`, and this scan never fires. `reentrant_instantiate_at_same_address_is_rejected` now passes through the nonce check. I'd remove the scan, or if you keep it, say it is only a backstop.

File: substrate/frame/revive/src/exec.rs [hunk 7]
```diff
@@ -2129,6 +2129,12 @@ where
 			self.exec_config,
 		)?;
 
+		// EIP-7702: the balance still moves, but a delegated EOA is never destroyed. Its
+		// `contract_info` is the authority's delegation trie and the target's code hash.
+		if AccountInfo::<T>::is_delegated(&contract_address) {
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:2134]: With the collision check in place this can no longer prevent a deletion. Authorizations are applied only before execution and skip contracts, and a delegated address can't be created anymore, so a delegated EOA is never in `contracts_created` and the `only_if_same_tx` filter at `exec.rs:1784` already skips it. The only observable effect left is that a SELFDESTRUCT in delegated code now gets the `Terminate { code_removed: false }` refund, while a regular contract that was not created in this transaction still pays the full cost for the same no-op.

I'd drop this together with `termination_of_delegated_eoa_created_in_same_tx_is_skipped`, which can only reach it by calling `set_delegation` from inside a constructor. If the cheaper charge is wanted, do it for every `only_if_same_tx` case in a separate PR.

File: substrate/frame/revive/src/tests/eip7702.rs [hunk 8]
```diff
@@ -2832,3 +2832,83 @@ fn self_delegation_traps_on_call() {
 		assert_err!(result.result, <Error<Test>>::ContractTrapped);
 	});
 }
+
+/// EIP-684: CREATE2 onto an address that holds a delegation entry must fail and leave the
+/// entry untouched, whether the delegation is active or already cleared.
+#[test]
+fn instantiate_onto_delegated_eoa_is_rejected() {
```
Comment (athei) [inline substrate/frame/revive/src/tests/eip7702.rs:2839]: These go through `bare_instantiate` on Solc only, so the path a factory contract takes (nested `push_frame`) is not covered on either backend. Please add a Solidity test in `tests/sol/contract.rs`, parametrized over Solc and Resolc like `root_call_can_create_and_destroy_in_next_block`: a factory doing `try new Child{salt: s}()` onto an address with an active delegation, a cleared delegation, nonce 1, and a funded address with nonce 0. Assert that the call doesn't revert, returns `address(0)` for the collisions and the create2 address for the funded case, that `AccountInfoOf` of the target is unchanged, and that the factory's nonce went up by one for each attempt.

File: substrate/frame/revive/src/lib.rs [hunk 9]
```diff
@@ -917,11 +917,14 @@ pub mod pallet {
 						};
 
 						let code_hash = *blob.code_hash();
-						let Ok(info) = <ContractInfo<T>>::new(&address, 0u32.into(), code_hash)
-							.inspect_err(|err| {
-								log::error!(target: LOG_TARGET, "Failed to create ContractInfo for {address:?}: {err:?}");
-							})
-						else {
+						let Ok(info) = <ContractInfo<T>>::new_without_collision_check(
```
Comment (athei) [inline substrate/frame/revive/src/lib.rs:920]: Nit: this also drops the duplicate guard. A genesis config that lists the same contract address twice used to log and skip the second entry. Now it overwrites `AccountInfoOf`, and because both entries derive the same trie id their storage merges. Writing the nonce after the contract is created (move the `frame_system::Account::mutate` below the match) lets genesis keep `ContractInfo::new`.

File: prdoc/pr_13202.prdoc [hunk 10]
```diff
@@ -0,0 +1,23 @@
+title: '[pallet-revive] Use an EIP-684 collision check for contract creation'
+doc:
+- audience: Runtime Dev
```
Comment (athei) [inline prdoc/pr_13202.prdoc:3]: This reads as review notes rather than a porting guide. Contract developers and tooling are affected too: CREATE/CREATE2 onto an address with a nonzero nonce or a delegation entry, active or cleared, now fails with `DuplicateContract` (the calling contract gets `address(0)`), so please add `Runtime User`. For `Runtime Dev`, say that out-of-tree callers of `ContractInfo::new` that install code on an existing account must switch to `new_without_collision_check`, and that genesis accounts with a nonce plus code and `eth_call` code overrides need no change. The `Fixes` link and the internals (`insert_contract`, `terminate_if_same_tx`) can go.

**Review by athei — CHANGES_REQUESTED** [review]
The collision check is the right fix, but EIP-684 is only half done: the creator's nonce is not consumed on a collision. Also blocking on the redeploy regression, the uncharged read in `is_create_collision`, and tests through a real factory contract.

The lazy precompile path in `new_frame` (`exec.rs:1147`, outside the diff) still calls `ContractInfo::new`. Creating info for a precompile is not a deployment, so it should use `new_without_collision_check`. Otherwise a precompile with contract info whose account has a nonce but no stored info fails every call with `DuplicateContract`.

File: substrate/frame/revive/src/storage.rs (hunk 4 above)
Comment (athei) [reply substrate/frame/revive/src/storage.rs]: Correction: "keyless fallback account" is wrong as a general statement. Fallback accounts are not keyless: Ethereum EOAs are fallback accounts, and every `AccountId32` controls the fallback of its own address through `dispatch_as_fallback_account`. For a contract address nobody controls the fallback account short of an address collision, so resetting the nonce in `do_terminate` is still safe.

File: substrate/frame/revive/src/storage.rs (hunk 5 above)
Comment (athei) [reply substrate/frame/revive/src/storage.rs]: Addendum: one H160 can have two System accounts with separate nonces, the mapped native account and its fallback (which the native account controls through `dispatch_as_fallback_account`). `to_account_id` only sees the native account when the address is mapped, `to_fallback_account_id` only sees the fallback. For a CREATE target the mapped case needs an address collision, so the fallback id is still the right choice, since that is where the contract lives. Please say in the doc comment that only the fallback account's nonce is checked.
